core/src/main/scala/cats/Fold.scala

```diff
@@ -0,0 +1,122 @@
+package cats
+
+/**
+ * Fold is designed to allow laziness/short-circuiting in foldLazy.
+ *
+ * It is a sum type that has three possible subtypes:
+ *
+ *   - Return(a): stop the fold with a value of `a`.
+ *   - Continue(f): continue the fold, suspending the computation `f`.
+ *   - Pass: continue the fold, with no additional computation.
+ *
+ * The meaning of these types can be made more clear with an example
+ * of the foldLazy method in action. Here's a method to count how many
+ * elements appear in a list before 3:
+ *
+ *     def f(n: Int): Fold[Int] =
+ *       if (n == 3) Fold.Return(0) else Fold.Continue(_ + 1)
+ *
+ *     val count: Lazy[Int] = List(1,2,3,4).foldLazy(Lazy(0))(f)
+ *
+ * When we call `count.force`, the following occurs:
+ *
+ *  - f(1) produces res0: Continue(_ + 1)
+ *  - f(2) produces res1: Continue(_ + 1)
+ *  - f(3) produces res2: Return(0)
+ *
+ * Now we unwind back through the continue instances:
+ *
+ *  - res2 produces 0
+ *  - res1(0) produces 1
+ *  - res0(1) produces 2
+ *
+ * And so the result is 2.
+ *
+ * This code searches an infinite stream for 77:
+ *
+ *    val found: Lazy[Boolean] =
```

Comment:
A bit pedantic, but if we wrap the docs here in {{{ ... }}} then ScalaDoc will treat it as preformatted text.